pkg/admission/clusterworkspaceshard/admission_test.go

```diff
@@ -309,8 +309,82 @@ func TestAdmit(t *testing.T) {
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			o := &clusterWorkspaceShard{
-				Handler:         admission.NewHandler(admission.Create, admission.Update),
-				externalAddress: "external.kcp.dev",
+				Handler:                 admission.NewHandler(admission.Create, admission.Update),
+				externalAddressProvider: func() string { return "external.kcp.dev" },
+			}
+			ctx := request.WithCluster(context.Background(), request.Cluster{Name: logicalcluster.New("root:org")})
+			if err := o.Admit(ctx, tt.a, nil); (err != nil) != tt.wantErr {
+				t.Fatalf("Validate() error = %v, wantErr %v", err, tt.wantErr)
+			} else if err == nil {
+				got, ok := tt.a.GetObject().(*unstructured.Unstructured)
+				require.True(t, ok, "expected unstructured, got %T", tt.a.GetObject())
+				expected := helpers.ToUnstructuredOrDie(tt.expectedObj)
+				if !apiequality.Semantic.DeepEqual(expected, got) {
+					t.Fatalf("unexpected result (A expected, B got): %s", diff.ObjectDiff(expected, got))
+				}
+			}
+		})
+	}
+}
+
+func TestAdmitWithoutExternalAddress(t *testing.T) {
```

Comment:
Do you think you could combine this into the previous test? Maybe add a bool to the test struct `provideExternalAddress` and have that set/not set `externalAddressProvider` in `o`?

Reply:
Good idea. I've added the plugin under test to the struct, so the test cases can still be run in parallel if we ever want it to.